File: scala/bob/src/main/scala/bob.scala
class Bob {
  def hey(input: String): String =
    if (isSilence(input))
      "Fine. Be that way!"
    else if (isShouting(input))
      "Whoa, chill out!"
    else if (isQuestion(input))
      "Sure."
    else
      "Whatever."

  private def isShouting(s: String) : Boolean =
    s.exists(Character.isUpperCase(_)) && !s.exists(Character.isLowerCase(_))

  private def isQuestion(s: String) : Boolean =
    s.endsWith("?")

  private def isSilence(s: String) : Boolean =
    s.trim.isEmpty
}

File: erlang/bob/bob.erl
-module(bob).
-export([response_for/1]).

response_for(S) ->
  case {is_question(S), is_shouting(S), is_silence(S)} of
    {_, _, true} -> "Fine. Be that way!";
    {_, true, _} -> "Whoa, chill out!";
    {true, _, _} -> "Sure.";
    _ -> "Whatever."
  end.

is_question(S) -> 
  case re:run(S, "\\?$") of
    {match, _} -> true;
    nomatch -> false
  end.

is_shouting(S) ->
  AtLeastOneUpperCase = "\\p{Lu}",
  case re:run(S, AtLeastOneUpperCase) of
    {match, _} -> S =:= string:to_upper(S);
    nomatch -> false
  end.

is_silence(S) -> [] =:= string:strip(S).
